### DISCUSSION TITLE
Automatic deserialization $type

### DISCUSSION THREAD

**Discussion by asapostolov** [ravendb/ravendb#13842] [Q&A] [answered]
I have this property in one of my classes (this is how it's stored in RavenDB)
``` JSON
"Settings": {
        "$type": "System.Collections.Generic.Dictionary`2[[System.String, mscorlib],[System.Object, mscorlib]], mscorlib",
        "lastUsedUnitId": {
              "1": "units-3368",
              "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib"
          }
}
```
The C# property looks like this

``` C#
public Dictionary<string, object> Settings { get; set; }
```

I assume (I think this is how this worked until recently) that `lastUsedUnitId` will automatically be deserialized to `Disctionary<int, string>`  by the RavenDB deserialization process because of the `$type` property there. However - I find it's deserialized as `JObject` instead which I have to convert manually in my code to `Dictionary`. 

Is there any way that this conversion can be done automatically? Am I missing something?

**Comment by maciejaszyk** [comment]
Hi, can you give us more information about which version of RavenDB Client and .NET you use?

I tried to reproduce it  ( https://gist.github.com/maciejaszyk/ca4fecd932c322dca787ba6d673752fb ), but the test works just fine.

**Reply by ayende** [reply]
To expand on what @maciejaszyk wrote, this _should_ work.
Check if you are doing anything in the conventions to modify the was serialization / deserialization works

**Reply by asapostolov** [reply]
This is very weird. Your example works fine, but I managed to create a failing test based on it when I copy-pasted my `lastUsedUnitId` value from a document from my server. The failing test changes the property with a script. Please check it out and let me know if it works on your end.

https://gist.github.com/asapostolov/faee1ffe9b09793940799b24729d8472

I should note that I've been using this functionality for years and it was working fine a couple of weeks ago (2-3). The only thing that might have changed in the meantime around it is the version of the RavenDB server. The above Gist with failing test is from 5.3.101 both Server & Client.

**Reply by ayende** [reply]
The issue is with likely because of the .NET version that you are running. When we save the document directly, we have:

`"System.Collections.Generic.Dictionary`2[[System.Int32, System.Private.CoreLib],[System.String, System.Private.CoreLib]], System.Private.CoreLib"`

When you do that via the patch, you have:

`"System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib"`

The problem is `mscorlib` vs. `System.Private.CoreLib` - are you using .NET framework and .NET Core in there together?

**Reply by asapostolov** [reply]
No, on my dev pc the strings are entirely the same (only difference is number of the `unit-` id), no conflict between the frameworks.

You can try to change the test I shared above to the right type and see if it works or you can run the test on a .NET Framework project to reproduce? (or both?)

**Reply by asapostolov** [reply]
Any progress about this?

**Reply by maciejaszyk** [reply]
Hi, 
I've created a ticket for this: https://issues.hibernatingrhinos.com/issue/RavenDB-18407
The test you provided does not pass on NET6 either. I will look for the cause of this.

Thanks!

**Reply by arekpalinski** [reply]
Hi @asapostolov,

As Oren pointed above the issue here is that your patch command explicitly uses `mscorlib` here: 

```
 ""$type"": ""System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib""
 ```
 
This is unknown library in .NET Core / .NET5+. Because of that when trying to deserialize such object with the usage of RavenDB Client in the app running on .NET Core / .NET5+ it fails so it fallbacks to `JObject`.
 
You mentioned that it was working a few weeks ago. What RavenDB version you were using then - both client and server?

**Reply by ayende** [reply]
Following up again, you are explicitly using `Patch` with a script here, or using `Patch` through the session API?

**Reply by ayende** [reply]
Here is a similar issue, see: https://issues.hibernatingrhinos.com/issue/RavenDB-18414

**Reply by asapostolov** [reply]
@arekpalinski 
> As Oren pointed above the issue here is that your patch command explicitly uses mscorlib

I addressed this concern in the discussion above - I'm not running the test (or my regular app) an .NET Core/.NET5+ - I'm running a .NET Framework 4.8 app  - you can make a .NET Framework project and try to run the test on it and confirm that it fails - it fails with me.

@ayende @arekpalinski  
This is a test I created in order to research why I was getting this error with later/est versions of Raven.Client. I'm not using `Patch` in the code of my app - I'm just loading an entity that was previously saved with that `$type` declaration. 

I downgraded Raven.Client in Friday to `5.2.2` (because of [this issue](https://github.com/ravendb/ravendb/discussions/13983)) and my app is working fine with that version - it desterilizes the property correctly into a Dictionary, BUT the test above fails (with `5.2.2`). I also tried the test with the previous stable version which I ran for a long time `5.1.4` and the test fails there too.

**Reply by asapostolov** [reply]
@arekpalinski 
Here's a gist with the same test failing for .NET 5

https://gist.github.com/asapostolov/01a5e95c0168065a091f4337da359671

**Reply by ayende** [reply]
thanks, looking into this

**Comment by ayende** [comment, marked as answer]
Okay, that took way too long to figure out, but now I have a good handle on what is going on here.

This will not work:

```
"Settings": {
        "$type": "System.Collections.Generic.Dictionary`2[[System.String, mscorlib],[System.Object, mscorlib]], mscorlib",
        "lastUsedUnitId": {
              "1": "units-3368",
              "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib"
          }
}
```


This, on the other hand, will:

```
"Settings": {
        "$type": "System.Collections.Generic.Dictionary`2[[System.String, mscorlib],[System.Object, mscorlib]], mscorlib",
        "lastUsedUnitId": {
              "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib"
             "1": "units-3368", 
          }
}
```

The key difference between those two, the `$type` position in the inner dictionary. 

`NewtonSoft.Json` requires that the "metadata" properties such as `$type` would be the first properties in the json in order to parse them properly.

**Reply by asapostolov** [reply]
Great, that's awesome that you found it!

So, the solution would be:
1. To take that into account when serializing the `$type` in the first place - I see that right now automatic serialization puts `$type` property last. (Is that managed by `Newtonsoft.Json` as well? If yes, would be quite inconsistent behavior)
```
  "lastUsedUnitId": {
            "1": "units-3368",
            "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib"
        }
```
2. To override the default behavior of `Newtonsoft.Json` in order to make it order-agnostic.

If it's option `1`, I'll need a script to update all the existing entities to move their type property as first in order? So for me probably option `2` is preferred.

What is strange is that if you remove the `Patch` operation from the test - the test passes, even when the `$type` is automatically serialized as the last json property. So there's difference when the doc is updated via `Raven.Client` vs when updated through `Patch`.

**Reply by ayende** [reply]
1. What automatic deserialization did that?
2. I doubt that this can work, would probably require patching JSON.Net

When you are working through the `Patch`, you are _manually_ controlling the order of fields.

**Reply by asapostolov** [reply]
1. The RavenDB default serialization puts `$type` last when saving the dictionary AND when loading - the deserialization into dictionary works as expected. If both are done right after the other. Check and debug `ShouldStoreAndRetrieveDictionary_Passes` in [this gist](https://gist.github.com/asapostolov/01a5e95c0168065a091f4337da359671)  and check the saved document. It looks like this in my server.

```JSON
{
    "$type": "TestRaven2.DeserializationTest+Test, TestRaven2",
    "Settings": {
        "$type": "System.Collections.Generic.Dictionary`2[[System.String, System.Private.CoreLib],[System.Object, System.Private.CoreLib]], System.Private.CoreLib",
        "lastUsedUnitId": {
            "1": "units-3368",
            "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, System.Private.CoreLib],[System.String, System.Private.CoreLib]], System.Private.CoreLib"
        }
    },
    "@metadata": {
        "@collection": "Tests",
        "Raven-Clr-Type": "TestRaven2.DeserializationTest+Test, TestRaven2"
    }
}
```

3. I'm not working through the `Patch` in my application code. I made the patch failing test in order to illustrate and communicate the issue with you guys. I'm working with entities that were created and are loaded via `Raven.Client`.

**Reply by ayende** [reply]
This looks like a bug in the studio, to be honest.
If you'll look at the raw results:

```
curl.exe http://127.0.0.1:57755/databases/ShouldStoreAndRetrieveDictionary_1/docs?id=tests%2F1-A | C:\Users\ayende\Downloads\jq.exe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   662    0   662    0     0  39125      0 --:--:-- --:--:-- --:--:-- 41375
{
  "Results": [
    {
      "$type": "SlowTests.Issues.RavenDB_18414_2+Test, SlowTests",
      "Settings": {
        "$type": "System.Collections.Generic.Dictionary`2[[System.String, System.Private.CoreLib],[System.Object, System.Private.CoreLib]], System.Private.CoreLib",
        "lastUsedUnitId": {
          "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, System.Private.CoreLib],[System.String, System.Private.CoreLib]], System.Private.CoreLib",
          "1": "units-3368"
        }
      },
      "@metadata": {
        "@collection": "Tests",
        "Raven-Clr-Type": "SlowTests.Issues.RavenDB_18414_2+Test, SlowTests",
        "@change-vector": "A:2-uamDUBQhbUmlXMjC+p6x5w",
        "@id": "tests/1-A",
        "@last-modified": "2022-04-04T14:20:48.2101267Z"
      }
    }
  ],
  "Includes": {}
}
```

The output is done properly. The Studio is showing the data in the wrong order for some reason (I assume that this is part of the formatting we do, will need to check).

However, editing this in the Studio would lead to the wrong sort order, which then cause the problem.

**Reply by ayende** [reply]
Confirmed, that is a bug, see: https://issues.hibernatingrhinos.com/issue/RavenDB-18417

**Reply by asapostolov** [reply]
> However, editing this in the Studio would lead to the wrong sort order, which then cause the problem.

Ok, so this is the issue and it's probably limited to a few documents that I've edited through the studio which messed up the ordering.

I'll do some more tests and confirm if it's working correctly in my application code as well.

**Reply by asapostolov** [reply]
I cannot fix my "broken" documents because there's auto-format "on save" in the studio. It would be ideal if the auto-format moves "$type" as first attribute by default.

**Reply by ayende** [reply]
You might be able to do that using `patch` operation, or just load and save the document through the API, no?

**Reply by asapostolov** [reply]
Yes, I can patch them 1-by-1

But there's something else - my app does not have this issue with v. 5.2.2 (it deserializes the 'broken' documents correctly) while the unit test crashes with 5.2.2. There might be 2 separate issues here, but I need to think of a way to reproduce the issue that my app is having with the versions above 5.2.3

**Reply by ayende** [reply]
This isn't related to the version of RavenDB. That hasn't changed. The issue is probably reordering from the Studio.
Once that happens, the rest follows.

**Comment by asapostolov** [comment]
@ayende 
I see that the issue we talked about is marked as fixed
https://issues.hibernatingrhinos.com/issue/RavenDB-18417

But the fix is very unhelpful.
![image](https://user-images.githubusercontent.com/1162471/169280301-aa83372d-a5b4-4c66-9a0d-04d73e4c19cd.png)
It just gives a useless warning that doesn't prevent the issue from occurring. 

Are you avoiding creating the fix of the JSON editor because it's too hard to change the algorithm for formatting? Or is there some other issue? I can imagine how easy might be to just **_skip_** the formatting function of the editor before saving when you detect that there is numeric key?

I also tried to fix the document with this script and it doesn't change the document if I don't change '-A' to '-a'.
``` Javascript
from Users
where id()='users-3073-b'
update {
    this.Settings["lastUsedUnitId"] = {
        "$type": "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.String, mscorlib]], mscorlib",
        "1": "units-14887-A"
    };
}
```
Which is extremely weird. Is that how patch is supposed to work? Not change the doc if the change is only in the ordering of the properties?

**Reply by ml054** [reply]
Hi @asapostolov 

Parsing raw document to json is not only used for formatting documents - we modify its content (metadata) so we want to operate on object literals not on strings. 

You may say: so let's write custom json serialize library to handle that case. It won't help us much, since then we break ECMA script standard (numeric props should go first). Also it might have negative performance impact (json.parse is native method). 

Using patch in studio executes that on server side, where we use javascript engine to patch objects. Running patch you described above will result in props order: (`1` and then `type`). 

You can observe same approach when using google chrome network tab: 
![image](https://user-images.githubusercontent.com/1250431/169476731-90d7242d-9a5b-4214-a3f5-33613e8b74e5.png)
it also sort property keys to match ecma script standard. 

I understand that this doesn't work well with your data structure, since serializer embeds `$type` property. 

So adding a warning about having numeric properties seems to be helpful - it allows you prevent issues and be aware of a weird part of emca script specs. Let's assume you want to use node.js client - then you also might end with this issue - so it is better to warns user asap. 

To fix your existing documents you can capture 'save document' request modify order of properties and issue this request manually via curl, postman etc.

**Reply by asapostolov** [reply]
Ok, I understand, thanks for explaining!

I think it's easier then, to rewrite my code in order to not use  `Dictionary<int, X>` in this collection. You might consider stating somewhere that you don't support `Dictionary<int, X>` with `$type` serialization in order to prevent this from happening in the future.

Thanks alot!

**Reply by ml054** [reply]
Hi, 

For following test:

```
 public class User
        {
            public string Id { get; set; }
            public string Name { get; set; }
            public IDictionary<string, string> TestingString { get; set; }
            public IDictionary<int, string> TestingInt { get; set; }
        }
        
        [Fact]
        public void CRUD_Operations()
        {
            using (var store = GetDocumentStore())
            {
                using (var newSession = store.OpenSession())
                {
                    newSession.Store(new User
                    {
                        Name = "user1",
                        TestingString = new Dictionary<string, string>() {
                            {"a", "b"}
                        },
                        TestingInt = new Dictionary<int, string>()
                        {
                            {5, "b"}
                        }
                    }, "users/1");

                    newSession.SaveChanges();
                }
            }
        }
```

We end up with document:
![image](https://user-images.githubusercontent.com/1250431/169820571-1a69874a-09f8-46f8-8252-0e0d9aec5d9a.png)


So we do support IDictionary. Problem starts with $type is introduced  +we use patch (see https://issues.hibernatingrhinos.com/issue/RavenDB-18655) or if you start playing with document in studio (see https://issues.hibernatingrhinos.com/issue/RavenDB-18417). 

That's the reason we warn about that.

**Reply by asapostolov** [reply]
Yes, you are right, thank you!
